File: cocos2dx-lua/poco/sdk/DefaultMatcher.lua
local EqualizationComparator = {}

function EqualizationComparator:compare(l, r)
    return l == r
end


local RegexpComparator = {}

function RegexpComparator:compare(origin, pattern)
    if origin == nil or pattern == nil then
        return false
    end
    return string.match(origin, pattern) ~= nil
end


local DefaultMatcher = {}
DefaultMatcher.__index = DefaultMatcher

DefaultMatcher.comparators = {
    ['attr='] = EqualizationComparator,
    ['attr.*='] = RegexpComparator,
}


function DefaultMatcher:match(cond, node)
    local op, args = unpack(cond)

    -- 条件匹配
    if op == 'or' then
        for _, arg in ipairs(args) do
            if self:match(arg, node) then
                return true
            end
        end
        return false
    end

    if op == 'and' then
        for _, arg in ipairs(args) do
            if not self:match(arg, node) then
                return false
            end
        end
        return true
    end

    -- 属性匹配
    comparator = self.comparators[op]
    if comparator ~= nil then
        local attribute, value = unpack(args)
        local targetValue = node:getAttr(attribute)
        return comparator:compare(targetValue, value) 
    end

    return false
end

return DefaultMatcher

File: cocos-creator/3X/sdk/DefaultMatcher.ts
interface IComparator {
    compare(cond: any, node: any): boolean
}

class EqualizationComparator implements IComparator {
    compare(cond: any, node: any): boolean {
        return cond === node
    }
}

class RegexpComparator implements IComparator {
    compare(origin: any, pattern: any): boolean {
        if (!origin || !pattern) {
            return false
        }
        return origin.toString().match(pattern) !== null
    }
}

export default class DefaultMatcher {
    comparators: { [key: string]: IComparator } = {
        "attr=": new EqualizationComparator(),
        "attr.*=": new RegexpComparator(),
    }
    match(cond: any, node: any): boolean {
        var op = cond[0]
        var args = cond[1]

        // 条件匹配
        if (op === "and") {
            for (var i in args) {
                var arg = args[i]
                if (!this.match(arg, node)) {
                    return false
                }
            }
            return true
        }

        if (op === "or") {
            for (var i in args) {
                var arg = args[i]
                if (this.match(arg, node)) {
                    return true
                }
            }
            return false
        }

        // 属性匹配
        var comparator: IComparator = this.comparators[op]
        if (comparator) {
            var attribute = args[0]
            var value = args[1]
            var targetValue = node.getAttr(attribute)
            return comparator.compare(targetValue, value)
        }

        return false
    }
}
